proposals/lambda-method-group-defaults.md

```diff
@@ -0,0 +1,324 @@
+```csharp
+void WriteInt(int i = 0) {
+  Console.Write(i);
+}
+
+var writeInt = WriteInt; // Inferred as Action<int>
+DoAction(writeInt, 3); // Ok, writeInt is an Action<int>
+
+void DoAction(Action<int> a, int p) {
+  a(p);
+}
+```
+Following this change, code of this nature would cease to compile.
+
+```csharp
+void WriteInt(int i = 0) {
+  Console.Write(i);
+}
+
+var writeInt = WriteInt; // Inferred as anonymous delegate type
+DoAction(writeInt, 3); // Error, cannot convert from anonymous delegate type to Action
+
+void DoAction(Action<int> a, int p) {
+  a(p);
+}
+```
+The impact of this breaking change needs to be considered. Fortunately, the use of `var` to infer the type of a method group has
+only been supported since C# 10, so only code which has been written since then which explicity relies on this behavior would break.
+
+## Detailed design
+
+[design]: #detailed-design
+
+### Grammar and Parser Changes
+This enhancement requires the following changes to the grammar for lambda expressions.
+```diff
+explicit_anonymous_function_parameter
+-    : anonymous_function_parameter_modifier? type identifier
++    : anonymous_function_parameter_modifier? type identifier default_argument?
+    ;
+```
+No changes to the grammar are necessary for method groups since this proposal would only change their semantics.
+
+### Binder Changes
+
+#### Synthesizing New Delegate Types
+As with the behavior for delegates with `ref` or `out` parameters, a new natural type is generated for each lambda or method group defined with any default parameter values.
+Note that in the below examples, the notation `a'`, `b'`, etc. is used to represent these anonymous delegate types.
+
+```csharp
+var addWithDefault = (int addTo = 2) => addTo + 1;
+// internal delegate int b'(int arg0 = 2);
+var printString = (string toPrint = "defaultString") => Console.WriteLine(toPrint);
+// internal delegate void c'(string arg0 = "defaultString");
+string pathJoin(string s1, string s2, string sep = "/") { return $"{s1}{sep}{s2}"; }
+var joinFunc = pathJoin;
+// internal delegate string d'(string arg0, string arg1, string arg3 = " ");
+```
+
+#### Conversion and Unification Behavior 
+The delegates described previously will be unified when the same parameter (based on position) has the same default value, regardless of parameter name.
+
+```csharp
+int E(int j = 13) {
+  return 11;
+}
+
+int F(int k = 0) {
+  return 3;
+}
+
+int G(int x = 13) {
+  return 4;
+}
+
+var a = (int i = 13) => 1;
+// internal delegate int b'(int arg0 = 13);
+var b = (int i = 0) => 2;
+// internal delegate int c'(int arg0 = 0);
+var c = (int i = 13) => 3;
+// internal delegate int b'(int arg0 = 13);
+var d = (int c = 13) => 1;
+// internal delegate int b'(int arg0 = 13);
+
+var e = E;
+// internal delegate int b'(int arg0 = 13);
+var f = F;
+// internal delegate int c'(int arg0 = 0);
+var g = G;
+// internal delegate int c'(int arg0 = 0);
```

Comment:
>c'(int arg0 = 0) [](http://example.com/codeflow?start=25&length=16)

Should this be `b'(int arg0 = 13)`?

Reply:
Yes it should be. Thanks for catching that